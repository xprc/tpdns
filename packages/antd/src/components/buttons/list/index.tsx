import React from "react";
import { Button } from "antd";
import { BarsOutlined } from "@ant-design/icons";
import {
    useCan,
    useNavigation,
    useTranslate,
    userFriendlyResourceName,
    useResource,
    useRouterContext,
} from "@pankod/refine-core";
import { RefineButtonTestIds } from "@pankod/refine-ui-types";

import { ListButtonProps } from "../types";

/**
 * `<ListButton>` is using Ant Design's {@link https://ant.design/components/button/ `<Button>`} component.
 * It uses the  {@link https://refine.dev/docs/core/hooks/navigation/useNavigation#list `list`} method from {@link https://refine.dev/docs/core/hooks/navigation/useNavigation `useNavigation`} under the hood.
 * It can be useful when redirecting the app to the list page route of resource}.
 *
 * @see {@link https://refine.dev/docs/ui-frameworks/antd/components/buttons/list-button} for more details.
 */
export const ListButton: React.FC<ListButtonProps> = ({
    resourceName: propResourceName,
    resourceNameOrRouteName: propResourceNameOrRouteName,
    hideText = false,
    accessControl,
    ignoreAccessControlProvider = false,
    children,
    onClick,
    ...rest
}) => {
    const accessControlEnabled =
        accessControl?.enabled ?? !ignoreAccessControlProvider;
    const hideIfUnauthorized = accessControl?.hideIfUnauthorized ?? false;
    const { listUrl: generateListUrl } = useNavigation();
    const { Link } = useRouterContext();
    const translate = useTranslate();

    const { resourceName, resource } = useResource({
        resourceName: propResourceName,
        resourceNameOrRouteName: propResourceNameOrRouteName,
    });

    const { data } = useCan({
        resource: resourceName,
        action: "list",
        queryOptions: {
            enabled: accessControlEnabled,
        },
        params: {
            resource,
        },
    });

    const createButtonDisabledTitle = () => {
        if (data?.can) return "";
        else if (data?.reason) return data.reason;
        else
            return translate(
                "buttons.notAccessTitle",
                "You don't have permission to access",
            );
    };

    const listUrl = generateListUrl(propResourceName ?? resource.route!);

    if (accessControlEnabled && hideIfUnauthorized && !data?.can) {
        return null;
    }

    return (
        <Link
            to={listUrl}
            replace={false}
            onClick={(e: React.MouseEvent<HTMLElement, MouseEvent>) => {
                if (data?.can === false) {
                    e.preventDefault();
                    return;
                }
                if (onClick) {
                    e.preventDefault();
                    onClick(e);
                }
            }}
        >
            <Button
                icon={<BarsOutlined />}
                disabled={data?.can === false}
                title={createButtonDisabledTitle()}
                data-testid={RefineButtonTestIds.ListButton}
                {...rest}
            >
                {!hideText &&
                    (children ??
                        translate(
                            `${resourceName}.titles.list`,
                            userFriendlyResourceName(
                                resource.label ?? resourceName,
                                "plural",
                            ),
                        ))}
            </Button>
        </Link>
    );
};
