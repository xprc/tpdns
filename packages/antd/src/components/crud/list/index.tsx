import React from "react";
import { PageHeader, Space } from "antd";
import {
    useResourceWithRoute,
    useRouterContext,
    useTranslate,
    userFriendlyResourceName,
    ResourceRouterParams,
    useRefineContext,
} from "@pankod/refine-core";

import { Breadcrumb, CreateButton } from "@components";
import { ListProps } from "../types";

/**
 * `<List>` provides us a layout for displaying the page.
 * It does not contain any logic but adds extra functionalities like a refresh button.
 *
 * @see {@link https://refine.dev/docs/ui-frameworks/antd/components/basic-views/list} for more details.
 */
export const List: React.FC<ListProps> = ({
    canCreate,
    title,
    children,
    createButtonProps,
    pageHeaderProps,
    resource: resourceFromProps,
    wrapperProps,
    contentProps,
    headerProps,
    breadcrumb: breadcrumbFromProps,
    headerButtonProps,
    headerButtons,
}) => {
    const { useParams } = useRouterContext();

    const { resource: routeResourceName } = useParams<ResourceRouterParams>();

    const translate = useTranslate();
    const resourceWithRoute = useResourceWithRoute();

    const resource = resourceWithRoute(resourceFromProps ?? routeResourceName);

    const isCreateButtonVisible =
        canCreate ?? (resource.canCreate || createButtonProps);

    const { options } = useRefineContext();
    const breadcrumb =
        typeof breadcrumbFromProps === "undefined"
            ? options?.breadcrumb
            : breadcrumbFromProps;

    const defaultExtra = isCreateButtonVisible ? (
        <CreateButton
            size="middle"
            resourceNameOrRouteName={resource.route}
            {...createButtonProps}
        />
    ) : null;

    return (
        <div {...(wrapperProps ?? {})}>
            <PageHeader
                ghost={false}
                title={
                    title ??
                    translate(
                        `${resource.name}.titles.list`,
                        userFriendlyResourceName(
                            resource.label ?? resource.name,
                            "plural",
                        ),
                    )
                }
                extra={
                    headerButtons ? (
                        <Space wrap {...headerButtonProps}>
                            {typeof headerButtons === "function"
                                ? headerButtons({
                                      defaultButtons: defaultExtra,
                                  })
                                : headerButtons}
                        </Space>
                    ) : (
                        defaultExtra
                    )
                }
                breadcrumb={
                    typeof breadcrumb !== "undefined" ? (
                        <>{breadcrumb}</> ?? undefined
                    ) : (
                        <Breadcrumb />
                    )
                }
                {...(pageHeaderProps ?? {})}
                {...(headerProps ?? {})}
            >
                <div {...(contentProps ?? {})}>{children}</div>
            </PageHeader>
        </div>
    );
};
