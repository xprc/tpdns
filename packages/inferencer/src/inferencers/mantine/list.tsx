import * as RefineMantine from "@pankod/refine-mantine";
import * as RefineReactTable from "@pankod/refine-react-table";

import { createInferencer } from "@/create-inferencer";
import {
    jsx,
    componentName,
    prettyString,
    accessor,
    printImports,
    dotAccessor,
    noOp,
    getVariableName,
} from "@/utilities";

import { ErrorComponent } from "./error";
import { LoadingComponent } from "./loading";
import { CodeViewerComponent } from "./code-viewer";

import { InferencerResultComponent, InferField } from "@/types";

const getAccessorKey = (field: InferField) => {
    return Array.isArray(field.accessor) || field.multiple
        ? `accessorKey: "${field.key}"`
        : field.accessor
        ? `accessorKey: "${dotAccessor(field.key, undefined, field.accessor)}"`
        : `accessorKey: "${field.key}"`;
};

/**
 * @experimental This is an experimental component
 */
export const ListInferencer: InferencerResultComponent = createInferencer({
    type: "list",
    additionalScope: [
        ["@pankod/refine-mantine", "RefineMantine", RefineMantine],
        ["@pankod/refine-react-table", "RefineReactTable", RefineReactTable],
    ],
    codeViewerComponent: CodeViewerComponent,
    loadingComponent: LoadingComponent,
    errorComponent: ErrorComponent,
    renderer: ({ resource, fields, isCustomPage }) => {
        const COMPONENT_NAME = componentName(
            resource.label ?? resource.name,
            "list",
        );
        const recordName = "tableData?.data";
        const imports: Array<[element: string, module: string]> = [
            ["IResourceComponentsProps", "@pankod/refine-core"],
            ["useTable", "@pankod/refine-react-table"],
            ["ColumnDef", "@pankod/refine-react-table"],
            ["flexRender", "@pankod/refine-react-table"],
            ["ScrollArea", "@pankod/refine-mantine"],
            ["List", "@pankod/refine-mantine"],
            ["Table", "@pankod/refine-mantine"],
            ["Pagination", "@pankod/refine-mantine"],
            ["Group", "@pankod/refine-mantine"],
            ["EditButton", "@pankod/refine-mantine"],
            ["ShowButton", "@pankod/refine-mantine"],
            ["DeleteButton", "@pankod/refine-mantine"],
        ];

        const relationFields: (InferField | null)[] = fields.filter(
            (field) => field?.relation && !field?.fieldable && field?.resource,
        );

        const relationHooksCode = relationFields
            .filter(Boolean)
            .map((field) => {
                if (field?.relation && !field.fieldable && field.resource) {
                    imports.push(["GetManyResponse", "@pankod/refine-core"]);
                    imports.push(["useMany", "@pankod/refine-core"]);

                    let idsString = "";

                    if (field.multiple) {
                        idsString = `[].concat(...(${recordName}?.map((item) => ${accessor(
                            "item",
                            field.key,
                            field.accessor,
                            false,
                        )}) ?? []))`;
                    } else {
                        idsString = `${recordName}?.map((item) => ${accessor(
                            "item",
                            field.key,
                            field.accessor,
                            false,
                        )}) ?? []`;
                    }

                    return `
                    const { data: ${getVariableName(field.key, "Data")} } =
                    useMany({
                        resource: "${field.resource.name}",
                        ids: ${idsString},
                        queryOptions: {
                            enabled: !!${recordName},
                        },
                    });
                    `;
                }
                return undefined;
            })
            .filter(Boolean);

        const relationVariableNames = relationFields
            ?.map((field) => {
                if (field && field.resource) {
                    return getVariableName(field.key, "Data");
                }
                return undefined;
            })
            .filter(Boolean);

        const renderRelationFields = (field: InferField) => {
            if (field.relation && field.resource) {
                const variableName = `${getVariableName(
                    field.key,
                    "Data",
                )}?.data`;

                if (Array.isArray(field.accessor)) {
                    // not handled - not possible case
                    return undefined;
                }

                const id = `id: "${field.key}"`;
                const header = `header: "${prettyString(field.key)}"`;
                const accessorKey = getAccessorKey(field);

                let cell = "";

                // if multiple, then map it with tagfield
                // if not, then just show the value

                if (field.multiple) {
                    imports.push(["TagField", "@pankod/refine-mantine"]);
                    let val = "item";

                    // for multiple
                    if (field?.relationInfer) {
                        val = accessor(
                            "item",
                            undefined,
                            field.relationInfer.accessor,
                        );
                    }

                    cell = `cell: function render({ getValue, table }) {
                        const meta = table.options.meta as {
                            ${getVariableName(
                                field.key,
                                "Data",
                            )}: GetManyResponse;
                        };

                        const ${getVariableName(
                            field.key,
                        )} = getValue<any[]>()?.map((item) => {
                            return meta.${getVariableName(
                                field.key,
                                "Data",
                            )}?.data?.find(
                                (resourceItems) => resourceItems.id === ${accessor(
                                    "item",
                                    undefined,
                                    field.accessor,
                                )}
                            );
                        })

                        return (
                            <Group spacing="xs">
                                {${getVariableName(
                                    field.key,
                                )}?.map((item, index) => (
                                    <TagField key={index} value={${val}} />
                                ))}
                            </Group>
                        )
                    }
                `;
                } else {
                    if (field?.relationInfer) {
                        cell = `cell: function render({ getValue, table }) {
                            const meta = table.options.meta as {
                                ${getVariableName(
                                    field.key,
                                    "Data",
                                )}: GetManyResponse;
                            };

                            const ${getVariableName(
                                field.key,
                            )} = meta.${variableName}?.find(
                                (item) => item.id === getValue<any>(),
                            );

                            return ${accessor(
                                getVariableName(field.key),
                                undefined,
                                field?.relationInfer?.accessor,
                            )} ?? "Loading...";
                        },`;
                    } else {
                        cell = "";
                    }
                }

                return `
                    {
                        ${id},
                        ${header},
                        ${accessorKey},
                        ${cell}
                    }
                `;
            }
            return undefined;
        };

        const imageFields = (field: InferField) => {
            if (field.type === "image") {
                imports.push(["Image", "@pankod/refine-mantine"]);

                const id = `id: "${field.key}"`;
                const accessorKey = getAccessorKey(field);
                const header = `header: "${prettyString(field.key)}"`;

                let cell = jsx`
                    cell: function render({ getValue }) {
                        ${field?.accessor ? "try {" : ""}
                            return <Image sx={{ maxWidth: "100px" }} src={${accessor(
                                "getValue<any>()",
                                undefined,
                                Array.isArray(field.accessor)
                                    ? field.accessor
                                    : undefined,
                                " + ",
                            )}} />
                        ${
                            field?.accessor
                                ? " } catch (error) { return null; }"
                                : ""
                        }
                       
                    }
                `;

                if (field.multiple) {
                    const val = accessor(
                        "item",
                        undefined,
                        field.accessor,
                        " + ",
                    );

                    cell = `
                        cell: function render({ getValue }) {
                            ${field?.accessor ? "try {" : ""}
                                return (
                                    <Group spacing="xs">
                                        {getValue<any[]>()?.map((item, index) => (
                                            <Image src={${val}} key={index} sx={{ maxWidth: "100px" }} />
                                        ))}
                                    </Group>
                                )
                            ${
                                field?.accessor
                                    ? " } catch (error) { return null; }"
                                    : ""
                            }
                        }
                    `;
                }

                return `
                    {
                        ${id},
                        ${accessorKey},
                        ${header},
                        ${cell}
                    }
                `;
            }
            return undefined;
        };

        const emailFields = (field: InferField) => {
            if (field.type === "email") {
                imports.push(["EmailField", "@pankod/refine-mantine"]);

                const id = `id: "${field.key}"`;
                const accessorKey = getAccessorKey(field);
                const header = `header: "${prettyString(field.key)}"`;

                let cell = jsx`
                    cell: function render({ getValue }) {
                        return <EmailField value={${accessor(
                            "getValue<any>()",
                            undefined,
                            Array.isArray(field.accessor)
                                ? field.accessor
                                : undefined,
                            ' + " " + ',
                        )}} />
                    }
                `;

                if (field.multiple) {
                    imports.push(["TagField", "@pankod/refine-mantine"]);

                    const val = accessor(
                        "item",
                        undefined,
                        field.accessor,
                        " + ",
                    );

                    cell = `
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<any>()?.map((item, index) => (
                                        <TagField value={${val}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `;
                }

                return `
                    {
                        ${id},
                        ${accessorKey},
                        ${header},
                        ${cell}
                    }
                `;
            }
            return undefined;
        };

        const urlFields = (field: InferField) => {
            if (field.type === "url") {
                imports.push(["UrlField", "@pankod/refine-mantine"]);

                const id = `id: "${field.key}"`;
                const accessorKey = getAccessorKey(field);
                const header = `header: "${prettyString(field.key)}"`;

                let cell = jsx`
                    cell: function render({ getValue }) {
                        return <UrlField value={${accessor(
                            "getValue<any>()",
                            undefined,
                            Array.isArray(field.accessor)
                                ? field.accessor
                                : undefined,
                            " + ",
                        )}} />
                    }
                `;

                if (field.multiple) {
                    imports.push(["TagField", "@pankod/refine-mantine"]);

                    const val = accessor(
                        "item",
                        undefined,
                        field.accessor,
                        " + ",
                    );

                    cell = `
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<any[]>()?.map((item, index) => (
                                        <TagField value={${val}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `;
                }

                return `
                    {
                        ${id},
                        ${accessorKey},
                        ${header},
                        ${cell}
                    }
                `;
            }
            return undefined;
        };

        const booleanFields = (field: InferField) => {
            if (field?.type === "boolean") {
                imports.push(["BooleanField", "@pankod/refine-mantine"]);

                const id = `id: "${field.key}"`;
                const accessorKey = getAccessorKey(field);
                const header = `header: "${prettyString(field.key)}"`;

                let cell = jsx`
                    cell: function render({ getValue }) {
                        return <BooleanField value={${accessor(
                            "getValue<any>()",
                            undefined,
                            Array.isArray(field.accessor)
                                ? field.accessor
                                : undefined,
                            " + ",
                        )}} />
                    }
                `;

                if (field.multiple) {
                    const val = accessor(
                        "item",
                        undefined,
                        field.accessor,
                        " + ",
                    );

                    cell = `
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<any[]>()?.map((item, index) => (
                                        <BooleanField value={${val}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `;
                }

                return `
                    {
                        ${id},
                        ${accessorKey},
                        ${header},
                        ${cell}
                    }
                `;
            }

            return undefined;
        };

        const dateFields = (field: InferField) => {
            if (field.type === "date") {
                imports.push(["DateField", "@pankod/refine-mantine"]);

                const id = `id: "${field.key}"`;
                const accessorKey = getAccessorKey(field);
                const header = `header: "${prettyString(field.key)}"`;

                let cell = jsx`
                    cell: function render({ getValue }) {
                        return <DateField value={${accessor(
                            "getValue<any>()",
                            undefined,
                            Array.isArray(field.accessor)
                                ? field.accessor
                                : undefined,
                            ' + " " + ',
                        )}} />
                    }
                `;

                if (field.multiple) {
                    const val = accessor(
                        "item",
                        undefined,
                        field.accessor,
                        " + ",
                    );

                    cell = `
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<any[]>()?.map((item, index) => (
                                        <DateField value={${val}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `;
                }

                return `
                    {
                        ${id},
                        ${accessorKey},
                        ${header},
                        ${cell}
                    }
                `;
            }
            return undefined;
        };

        const richtextFields = (field: InferField) => {
            if (field?.type === "richtext") {
                imports.push(["MarkdownField", "@pankod/refine-mantine"]);

                const id = `id: "${field.key}"`;
                const accessorKey = getAccessorKey(field);
                const header = `header: "${prettyString(field.key)}"`;

                let cell = jsx`
                    cell: function render({ getValue }) {
                        return <MarkdownField value={(${accessor(
                            "getValue<string>()",
                            undefined,
                            Array.isArray(field.accessor)
                                ? field.accessor
                                : undefined,
                        )})?.slice(0, 80) + "..." } />
                    }
                `;

                if (field.multiple) {
                    const val = accessor(
                        "item",
                        undefined,
                        field.accessor,
                        " + ",
                    );

                    cell = `
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<string[]>()?.map((item, index) => (
                                        <MarkdownField value={${val}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `;
                }

                return `
                    {
                        ${id},
                        ${accessorKey},
                        ${header},
                        ${cell}
                    }
                `;
            }

            return undefined;
        };

        const basicFields = (field: InferField) => {
            if (field && (field.type === "text" || field.type === "number")) {
                const id = `id: "${field.key}"`;
                const accessorKey = getAccessorKey(field);
                const header = `header: "${prettyString(field.key)}"`;

                let cell = "";

                if (field.multiple) {
                    imports.push(["TagField", "@pankod/refine-mantine"]);

                    const val = accessor(
                        "item",
                        undefined,
                        field.accessor,
                        ' + " " + ',
                    );

                    cell = `
                        cell: function render({ getValue }) {
                            return (
                                <Group spacing="xs">
                                    {getValue<any[]>()?.map((item, index) => (
                                        <TagField value={${val}} key={index} />
                                    ))}
                                </Group>
                            )
                        }
                    `;
                }

                if (!field.multiple && Array.isArray(field.accessor)) {
                    cell = `
                        cell: function render({ getValue }) {
                            return (
                                <>{${accessor(
                                    "getValue()",
                                    field.key,
                                    field.accessor,
                                )}}</>
                            );
                        }
                    `;
                }

                return `
                    {
                        ${id},
                        ${accessorKey},
                        ${header},
                        ${cell}
                    }
                `;
            }
            return undefined;
        };

        const { canEdit, canShow, canDelete } = resource ?? {};

        if (canEdit) {
            imports.push(["EditButton", "@pankod/refine-mantine"]);
        }
        if (canShow) {
            imports.push(["ShowButton", "@pankod/refine-mantine"]);
        }
        if (canDelete) {
            imports.push(["DeleteButton", "@pankod/refine-mantine"]);
        }

        const actionButtons =
            canEdit || canShow || canDelete
                ? jsx`
        {
            id: "actions",
            accessorKey: "id",
            header: "Actions",
            cell: function render({ getValue }) {
                return (
                    <Group spacing="xs" noWrap>
                        ${
                            canShow
                                ? jsx`
                        <ShowButton
                            hideText
                            recordItemId={getValue() as string}
                        />
                        `
                                : ""
                        }
                        ${
                            canEdit
                                ? jsx`<EditButton
                            hideText
                            recordItemId={getValue() as string}
                        />`
                                : ""
                        }
                        ${
                            canDelete
                                ? jsx`
                        <DeleteButton
                            hideText
                            recordItemId={getValue() as string}
                        />
                        `
                                : ""
                        }
                    </Group>
                );
            },
        },
            `
                : "";

        const renderedFields: Array<string | undefined> = fields.map(
            (field) => {
                switch (field?.type) {
                    case "text":
                    case "number":
                        return basicFields(field);
                    case "richtext":
                        return richtextFields(field);
                    case "email":
                        return emailFields(field);
                    case "image":
                        return imageFields(field);
                    case "date":
                        return dateFields(field);
                    case "boolean":
                        return booleanFields(field);
                    case "url":
                        return urlFields(field);
                    case "relation":
                        return renderRelationFields(field);
                    default:
                        return undefined;
                }
            },
        );

        noOp(imports);

        return jsx`
        import React from "react";
        ${printImports(imports)}
        
        export const ${COMPONENT_NAME}: React.FC<IResourceComponentsProps> = () => {
            const columns = React.useMemo<ColumnDef<any>[]>(() => [
                ${[...renderedFields, actionButtons].filter(Boolean).join(",")}
            ], []);

            const {
                getHeaderGroups,
                getRowModel,
                setOptions,
                refineCore: {
                    setCurrent,
                    pageCount,
                    current,
                    tableQueryResult: { data: tableData },
                },
            } = useTable({
                columns,
                ${
                    isCustomPage
                        ? `
                refineCoreProps: {
                    resource: "${resource.name}",
                }
                `
                        : ""
                }
            });

            ${relationHooksCode}

            setOptions((prev) => ({
                ...prev,
                meta: {
                    ...prev.meta,
                    ${relationVariableNames.join(", ")}
                },
            }));

            return (
                <List>
                    <ScrollArea>
                        <Table highlightOnHover>
                            <thead>
                                {getHeaderGroups().map((headerGroup) => (
                                    <tr key={headerGroup.id}>
                                        {headerGroup.headers.map((header) => {
                                            return (
                                                <th key={header.id}>
                                                    {!header.isPlaceholder && (
                                                        flexRender(
                                                            header.column
                                                                .columnDef
                                                                .header,
                                                            header.getContext(),
                                                        )
                                                    )}
                                                </th>
                                            );
                                        })}
                                    </tr>
                                ))}
                            </thead>
                            <tbody>
                                {getRowModel().rows.map((row) => {
                                    return (
                                        <tr key={row.id}>
                                            {row.getVisibleCells().map((cell) => {
                                                return (
                                                    <td key={cell.id}>
                                                        {flexRender(
                                                            cell.column.columnDef.cell,
                                                            cell.getContext(),
                                                        )}
                                                    </td>
                                                );
                                            })}
                                        </tr>
                                    );
                                })}
                            </tbody>
                        </Table>
                    </ScrollArea>    
                    <br />
                    <Pagination
                        position="right"
                        total={pageCount}
                        page={current}
                        onChange={setCurrent}
                    />
                </List>
            );
        };
        `;
    },
});
