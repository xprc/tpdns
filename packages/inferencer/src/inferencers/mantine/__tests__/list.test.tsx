import React from "react";
import { Route, Routes } from "react-router-dom";

import { render, act, TestWrapper } from "@test";
import { ListInferencer } from "../list";

describe("MantineListInferencer", () => {
    it("should match the snapshot", async () => {
        const Wrapper = TestWrapper({
            routerInitialEntries: ["/posts"],
            resources: [
                {
                    name: "posts",
                    list: ListInferencer,
                },
                {
                    name: "categories",
                },
                {
                    name: "tags",
                },
                { name: "users" },
            ],
        });

        const rendering = render(
            <Wrapper>
                <Routes>
                    <Route
                        path="/:resource"
                        element={<ListInferencer resource="posts" />}
                    />
                </Routes>
            </Wrapper>,
        );

        await act(async () => {
            await new Promise((resolve) => setTimeout(resolve, 2000));
        });
        await act(async () => {
            await new Promise((resolve) => setTimeout(resolve, 2000));
        });
        await act(async () => {
            await new Promise((resolve) => setTimeout(resolve, 2000));
        });
        await act(async () => {
            await new Promise((resolve) => setTimeout(resolve, 2000));
        });

        const node = rendering.asFragment();

        expect(node).toMatchSnapshot();
    });
});
