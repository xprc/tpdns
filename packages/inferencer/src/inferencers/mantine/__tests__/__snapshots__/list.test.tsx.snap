// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MantineListInferencer should match the snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="mantine-Paper-root mantine-Card-root mantine-xoj88x"
    >
      <div
        class="mantine-Group-root mantine-1dn2tzg"
      >
        <div
          class="mantine-Stack-root mantine-1xft3nm"
        >
          <h3
            class="mantine-Text-root mantine-Title-root mantine-6hw97m"
          >
            Posts
          </h3>
        </div>
        <div
          class="mantine-Group-root mantine-1g4q40w"
        />
      </div>
      <div
        class="mantine-1r0zdw0"
      >
        <div
          class="mantine-ScrollArea-root mantine-jghxib"
          dir="ltr"
          style="position: relative; --radix-scroll-area-corner-width: 0px; --radix-scroll-area-corner-height: 0px;"
        >
          <style>
            [data-radix-scroll-area-viewport]{scrollbar-width:none;-ms-overflow-style:none;-webkit-overflow-scrolling:touch;}[data-radix-scroll-area-viewport]::-webkit-scrollbar{display:none}
          </style>
          <div
            class="mantine-xlwgkm mantine-ScrollArea-viewport"
            data-radix-scroll-area-viewport=""
            style="overflow-x: scroll; overflow-y: scroll;"
          >
            <div
              style="min-width: 100%; display: table;"
            >
              <table
                class="mantine-Table-root mantine-1n3n130"
                data-hover="true"
              >
                <thead>
                  <tr>
                    <th>
                      Id
                    </th>
                    <th>
                      Title
                    </th>
                    <th>
                      Slug
                    </th>
                    <th>
                      Content
                    </th>
                    <th>
                      Hit
                    </th>
                    <th>
                      Category
                    </th>
                    <th>
                      User
                    </th>
                    <th>
                      Status
                    </th>
                    <th>
                      Status Color
                    </th>
                    <th>
                      Created At
                    </th>
                    <th>
                      Published At
                    </th>
                    <th>
                      Image
                    </th>
                    <th>
                      Tags
                    </th>
                    <th>
                      Language
                    </th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td>
                      11
                    </td>
                    <td>
                      Nobis exercitationem officia quia est.
                    </td>
                    <td>
                      provident-culpa-facere
                    </td>
                    <td>
                      <p>
                        Non velit blanditiis optio quo quaerat nam aperiam. Odit tempore consequatur vol...
                      </p>
                    </td>
                    <td>
                      920557
                    </td>
                    <td>
                      Numquam Saepe Illo
                    </td>
                    <td>
                      Eriberto Leannon
                    </td>
                    <td>
                      draft
                    </td>
                    <td>
                      orange
                    </td>
                    <td>
                      <div
                        class="mantine-Text-root mantine-1pbxw0k"
                      >
                        04/02/2022
                      </div>
                    </td>
                    <td>
                      <div
                        class="mantine-Text-root mantine-1pbxw0k"
                      >
                        08/10/2021
                      </div>
                    </td>
                    <td>
                      <div
                        class="mantine-Group-root mantine-1g4q40w"
                      >
                        <div
                          class="mantine-Image-root mantine-r5ruyt"
                          style="width: 100%;"
                        >
                          <figure
                            class="mantine-11nhzn5 mantine-Image-figure"
                          >
                            <div
                              class="mantine-qqmv3w mantine-Image-imageWrapper"
                            >
                              <img
                                class="mantine-3y8yz3 mantine-Image-image"
                                src="http://loremflickr.com/640/480/abstract"
                                style="object-fit: cover; width: 100%; height: auto;"
                              />
                            </div>
                          </figure>
                        </div>
                      </div>
                    </td>
                    <td>
                      <div
                        class="mantine-Group-root mantine-1g4q40w"
                      >
                        <div
                          class="mantine-Chip-root mantine-15po0m8"
                        >
                          <input
                            class="mantine-106387w mantine-Chip-input"
                            id="mantine-r2"
                            type="checkbox"
                            value=""
                          />
                          <label
                            class="__mantine-ref-label mantine-543i07 mantine-Chip-label"
                            data-variant="outline"
                            for="mantine-r2"
                          >
                            Nobis Et Ut
                          </label>
                        </div>
                      </div>
                    </td>
                    <td>
                      2
                    </td>
                  </tr>
                  <tr>
                    <td>
                      21
                    </td>
                    <td>
                      Eaque in nesciunt.
                    </td>
                    <td>
                      quod-quos-commodi
                    </td>
                    <td>
                      <p>
                        Ut velit quo. Sed beatae et sit ea. Perferendis pariatur doloremque deleniti sap...
                      </p>
                    </td>
                    <td>
                      175149
                    </td>
                    <td>
                      Sit Natus Voluptatibus
                    </td>
                    <td>
                      Tressie Reichel
                    </td>
                    <td>
                      published
                    </td>
                    <td>
                      lime
                    </td>
                    <td>
                      <div
                        class="mantine-Text-root mantine-1pbxw0k"
                      >
                        09/24/2022
                      </div>
                    </td>
                    <td>
                      <div
                        class="mantine-Text-root mantine-1pbxw0k"
                      >
                        09/01/2021
                      </div>
                    </td>
                    <td>
                      <div
                        class="mantine-Group-root mantine-1g4q40w"
                      >
                        <div
                          class="mantine-Image-root mantine-r5ruyt"
                          style="width: 100%;"
                        >
                          <figure
                            class="mantine-11nhzn5 mantine-Image-figure"
                          >
                            <div
                              class="mantine-qqmv3w mantine-Image-imageWrapper"
                            >
                              <img
                                class="mantine-3y8yz3 mantine-Image-image"
                                src="http://loremflickr.com/640/480/abstract"
                                style="object-fit: cover; width: 100%; height: auto;"
                              />
                            </div>
                          </figure>
                        </div>
                      </div>
                    </td>
                    <td>
                      <div
                        class="mantine-Group-root mantine-1g4q40w"
                      >
                        <div
                          class="mantine-Chip-root mantine-15po0m8"
                        >
                          <input
                            class="mantine-106387w mantine-Chip-input"
                            id="mantine-r3"
                            type="checkbox"
                            value=""
                          />
                          <label
                            class="__mantine-ref-label mantine-543i07 mantine-Chip-label"
                            data-variant="outline"
                            for="mantine-r3"
                          >
                            Error Eos Et
                          </label>
                        </div>
                        <div
                          class="mantine-Chip-root mantine-15po0m8"
                        >
                          <input
                            class="mantine-106387w mantine-Chip-input"
                            id="mantine-r4"
                            type="checkbox"
                            value=""
                          />
                          <label
                            class="__mantine-ref-label mantine-543i07 mantine-Chip-label"
                            data-variant="outline"
                            for="mantine-r4"
                          >
                            Nobis Et Ut
                          </label>
                        </div>
                      </div>
                    </td>
                    <td>
                      2
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
        <br />
        <div
          class="mantine-Group-root mantine-uemwr9"
          role="navigation"
        >
          <button
            aria-disabled="true"
            class="mantine-109t3tn mantine-Pagination-item"
            disabled=""
            type="button"
          >
            <svg
              fill="none"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M7.219 8l3.3 3.3-.943.943L5.333 8l4.243-4.243.943.943-3.3 3.3z"
                fill="currentColor"
              />
            </svg>
          </button>
          <button
            aria-current="page"
            class="mantine-109t3tn mantine-Pagination-item"
            data-active="true"
            tabindex="0"
            type="button"
          >
            1
          </button>
          <button
            aria-disabled="true"
            class="mantine-109t3tn mantine-Pagination-item"
            disabled=""
            type="button"
          >
            <svg
              fill="none"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8.781 8l-3.3-3.3.943-.943L10.667 8l-4.243 4.243-.943-.943 3.3-3.3z"
                fill="currentColor"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    style="height: 50px;"
  />
</DocumentFragment>
`;
